### ISSUE TITLE
How to use this solver to solve general optimization problem?

### ISSUE THREAD

**Issue by NengQian** [fabiodimarco/torch-levenberg-marquardt#2]
Hi
This work looks excellent. 
But how to use this to solve a general optimzaiton instead of training a neural network?
For example how to use this solver to solve a problem like Chained Rosenbrock function which people do in [levmar ](https://users.ics.forth.gr/~lourakis/levmar/)？

**Comment by NengQian** [comment]
I think I may know how. Maybe just define my own nn module like in this page. It not really need to be a neural network.
https://pytorch.org/tutorials/beginner/examples_nn/polynomial_module.html

**Comment by fabiodimarco** [comment]
Hi @NengQian, you’re on the right track. You can create a custom `torch.nn.Module` in which you define the trainable parameters in the `__init__` method and specify how they’re used in the `forward` method to produce the outputs.

In the example below, the forward method returns the values before they’re squared and summed, allowing the MSE loss to take care of those operations. I set y_dummy to zero because the MSE loss effectively computes $(outputs - 0)^2$ in this context.

This is the equation I’ve been using (though I’m not entirely sure it’s exactly what you need):

$$
f(\mathbf{x}) = \sum_{i=1}^{n-1} \left[ 100 \cdot \left( x_{i+1} - x_i^2 \right)^2 + \left( x_i - 1 \right)^2 \right]
$$

Below is a possible implementation. I’ve defined a dummy dataset because my fit function expects one by default, but you could remove it if you write your own custom training loop.

```
# %%
# !%reload_ext autoreload
# !%autoreload 2

import time

import torch
import torch.nn as nn
import torch_levenberg_marquardt as tlm
from torch.utils.data import TensorDataset

# Set PyTorch to use high precision for matrix multiplication
torch.set_float32_matmul_precision('high')

# Detect CUDA device for acceleration
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
print(f'Using device: {device}')


# %%
# Define the Chained Rosenbrock model
class ChainedRosenbrock(nn.Module):
    def __init__(self, n: int) -> None:
        super().__init__()
        self.x = nn.Parameter(torch.randn(n))

    def forward(self, _=None) -> torch.Tensor:
        differences = self.x[1:] - self.x[:-1] ** 2
        deviations = self.x[:-1] - 1
        return torch.cat((differences, deviations))


# %%
# Create a dummy dataset
n = 5  # Dimension of x
target_size = 2 * (n - 1)

# Dummy 'input' for DataLoader (unused by the model)
x_dummy = torch.zeros(target_size, dtype=torch.float32).unsqueeze(0).to(device)

# Zero vector used as target for the MSE Loss
y_target = torch.zeros(target_size, dtype=torch.float32).unsqueeze(0).to(device)

dataset = TensorDataset(x_dummy, y_target)
train_loader = tlm.utils.FastDataLoader(
    dataset=dataset,
    batch_size=1,
    shuffle=False,
    device=device,
)

# %%
# Define two models: one for Adam, one for LM
model = ChainedRosenbrock(n).to(device)
model_lm = ChainedRosenbrock(n).to(device)

num_parameters = sum(p.numel() for p in model.parameters() if p.requires_grad)
print(f'Number of trainable parameters: {num_parameters}')

module = tlm.training.OptimizerModule(
    model=model,
    optimizer=torch.optim.Adam(model.parameters(), lr=0.1),
    loss_fn=nn.MSELoss(),
)

module_lm = tlm.training.LevenbergMarquardtModule(
    model=model_lm,
    loss_fn=tlm.loss.MSELoss(),
    learning_rate=0.1,
    attempts_per_step=10,
    solve_method='qr',
    damping_strategy=tlm.damping.StandardDampingStrategy(damping_mode='adaptive'),
)

# %%
# Train model
print('Training with Adam optimizer...')
t1_start = time.perf_counter()
tlm.utils.fit(module, train_loader, epochs=200)
t1_stop = time.perf_counter()
print(f'Training (Adam) completed. Elapsed time: {t1_stop - t1_start:.2f} seconds\n')

print('Training with Levenberg-Marquardt...')
t2_start = time.perf_counter()
tlm.utils.fit(module_lm, train_loader, epochs=200)
t2_stop = time.perf_counter()
print(f'Training (LM) completed. Elapsed time: {t2_stop - t2_start:.2f} seconds\n')

```

I’m not certain this is fully correct, but it should give you a good starting point. Also, if you don’t mind, could you provide some examples of plots that might be useful for visualizing the results of this problem? I’m not familiar with it, but it could be useful to add an notebook in the examples folder for this problem.

**Comment by nengqian11** [comment]
Hi @fabiodimarco . Thanks for your reply! This is very helpful.
I can run the script now.
To be honset, I am also not familiar with this problem. The reason I asked about this problem is because before I was using the levmar cpp lib to solve my optimization problem, and the  Chained Rosenbrock function is a benchmark problem (check [here)](https://github.com/jturney/levmar/blob/master/src/lmdemo.c).
Thus I actually want to evaluate your implementation versus the levmar lib.
But ye, let me first try more with your implementataion, and then see what I can help you.
